File: java/Apple/Arrays_and_Strings/Two_Sum.java
package Apple.Arrays_and_Strings;
import java.util.Arrays;
import java.util.Map;
import java.util.HashMap;
public class Two_Sum {
    // brute force
    public int[] TwoSum(int[] nums, int target) {
        for (int i = 0; i < nums.length; ++i) {
            for (int j = i + 1; j < nums.length; ++j)
                if (nums[i] + nums[j] == target)
                    return new int[] { i, j };
        }
        return new int[] {};
    }

    public int[] TwoSum2(int[] nums, int target) {
            Map<Integer, Integer> res = new HashMap<>();
            for (int i=0;i<nums.length;++i){
                if (res.containsKey(nums[i])) return new int[] {i,res.get(nums[i])};
            
                res.put(target-nums[i],i);
            }
            return new int[] {};
            
    }

    public static void main(String[] args) {
        Two_Sum sol = new Two_Sum();
        // test case
        int[] nums = { 2, 7, 11, 15 };
        int target = 9;
        System.out.println(Arrays.toString(sol.TwoSum2(nums, target)));
    }
}//

File: py/Apple/Arrays_and_Strings/Two_Sum.py
# brute force
# speed : O(n^2)
# space : O(1)


def twoSum(nums, target):
    for i in range(len(nums)):
        for j in range(i+1, len(nums)):
            if nums[i]+nums[j] == target:
                return [i, j]

# hashmap
# speed : O(n)
# space : O(n)


def twoSum2(nums, target):
    res = {}
    for i in range(len(nums)):
        if nums[i] in res:
            return [i, res[nums[i]]]
        else:
            res[target-nums[i]] = i


if __name__ == '__main__':
    nums = [2, 7, 11, 15]
    target = 9
    print(twoSum2(nums, target))
